### ISSUE TITLE
``ed.copy`` not replacing all ancestors

### ISSUE THREAD

**Issue by bayerj** [blei-lab/edward#798]
**tl;dr** `ed.copy` seems to have an issue with recursion, yielding results different to what the docs suggest and a user would, imho, expect. A minimal example can be found here: https://gist.github.com/bayerj/c1a9e684ad0e3b9b2fe1dcdb6976c688.

----

Assume the following PGM:

```
z[0] ---> z[1] ---> z[2] ---> z[3] ----> z[4]
```

I then extend this to the following:

```
z[0]  ---> z[1]  ---> z[2]  ---> z[3]  ---> z[4]
             |         |           |         |
             V         V           V         V
qz[0]      qz[1]      qz[2]      qz[3]      qz[4]
```

Now I want a PGM where all occurences of `z[i]` are replaced by `qz[i]`. This is a use case in `KLqp` for sequential models.

If I call `ed.copy` in this case, I would expect to end up with

```
qz[0] ---> qz[1] ---> qz[2] ---> qz[3] ----> qz[4]
```

However, "something else" happens. I will illustrate in the case of just copying `z[4]` where `z[:4]` is replaced with `qz[:4]`, i.e. running

```ed.copy(z[4], dict(zip(zs, qzs)))``` 

What I then end up with is effectively:

```
z[0] -> z[1] -> z[2] -> qz[3] ---> copy of z4
```

Long story short, replacings seem only to happen up to the first level. I guess the recursion of `ed.copy` is doing something wrong. I went into the code and [line 229 in ed.copy](https://github.com/blei-lab/edward/blob/master/edward/util/random_variables.py#L229) looked suspicious: the result of the recursive `ed.copy` is not used anywhere.

Since `ed.copy` is quite involved, I wonder if someone could assist me in looking into this further. Especially, whether my assumption is right that `ed.copy` does not mutate graphs, but always creates a new one.

**Comment by bkayalibay** [comment]
For this to work, one needs to call `copy` as `copy(x, latent_vars, scope='foo', copy_q=True, replace_itself=True)`. In the following, I'll try to go through the code of `copy` to show why one needs the arguments `copy_q` and `replace_itself` to be `True`. It turned out to be a rather lengthy text, so I wrote [this short](https://gist.github.com/bkayalibay/c2cd859b7a3bbc6515981120b5de6e5b) script to illustrate the effect of these two arguments, based on a slightly modified version of @bayerj's script.

I think the relevant part of `copy` to us is lines 231-250 in `edward/util/random_variables.py` :

```python
if isinstance(org_instance, RandomVariable):
  rv = org_instance

  # If it has copiable arguments, copy them.
  args = [_copy_default(arg, dict_swap, scope, True, copy_q, False)
              for arg in rv._args]

  kwargs = {}
  for key, value in six.iteritems(rv._kwargs):
    if isinstance(value, list):
      kwargs[key] = [_copy_default(v, dict_swap, scope, True, copy_q, False)
                             for v in value]
    else:
      kwargs[key] = _copy_default(
        value, dict_swap, scope, True, copy_q, False)
  kwargs['name'] = new_name

  # Create new random variable with copied arguments.
  try:
    new_rv = type(rv)(*args, **kwargs)
```

Here a new random variable is created by recursively copying the arguments of the old one. In our case these are:

```python
In [5]: aggregated._kwargs
Out[5]: 
{'loc': <ed.RandomVariable 'z3/' shape=(1,) dtype=float32>,
 'name': 'z4/',
 'scale': 1.0}
```

So now z3 is going to get copied. The difference to the original call of copy is that this time function argument `replace_itself` is set to `True` so when execution reaches the following point:

```python
if org_instance in dict_swap and replace_itself:
  org_instance = dict_swap[org_instance]
  if not copy_q:
    return org_instance
```

`z3` will be replaced by `qz3` and the recursion will end here unless `copy_q` is set to `True`. If `copy_q` is set, execution will recurse one level deeper into the arguments of `qz3` which are:

```python
In [17]: qzs[3]._kwargs
Out[17]: 
{'loc': <ed.RandomVariable 'z2/' shape=(1,) dtype=float32>,
 'name': 'qz3/',
 'scale': 1.0}
```

And so on.

Our computational graph looks like this:

```
zs[0] -> zs[1] -> zs[2] -> zs[3] -> zs[4]
  |        |        |        |
  v        v        v        v
qzs[1]   qzs[2]   qzs[3]   qzs[4]
```

Where a downward arrow and a side-ways arrow can, in general, be two different types of computation.  These random variables are paired as:

```
 zs[0]   zs[1]   zs[2]   zs[3]   zs[4]
   |       |       |       |       |
qzs[0]  qzs[1]  qzs[2]  qzs[3]  qzs[4]
```

We've seen that `copy(zs[4], latent_vars, copy_q=False)` will look at the arguments of `zs[4]`, which are, in general, going to be functions of `zs[3]` but for the sake of simplicity let's assume the mean of `zs[4]` is just a sample from `zs[3]`, i.e. `zs[3]` itself. In that case, since `zs[3]` is in the swap dictionary, it will simply be replaced by `qzs[3]`. Remember that `qzs[3]` is obtained by taking the downward arrow from `zs[2]`. The important thing here is that  the copy of `zs[4]` is not going to be equal to `downward_arrow(qzs[3])`. Instead, it will be equal to what you get if you apply the "sideways arrow" to `qzs[3]`, which in general is a different computation than the "downward arrow".  I'll illustrate what I mean with an example:

In the toy example of @bayerj , both arrows behave the same way. `a -> b` means: `b = Normal(a, scale=1.0)` and the downward arrow has the same meaning. Let's modify this so that the arrows have different meanings: `b = Normal(a * phi, scale=1.0)` where `phi=1` for the sideways arrow and `phi=2` for the downward arrow.

In this case, after calling `copy(zs[4], latent_vars)` we want to get the result of the following statement (not taking `copy_q` into account and just assuming everything works ideally):

```
return Normal(qz3 * 2, scale=1)
 where
  qz3 = Normal(qz2 * 2, scale=1)
  qz2 = Normal(qz1 * 2, scale=1)
  qz1 = Normal(qz0 * 2, scale=1)
```

In real life, if we set `copy_q=False`, we will get this:

```
# assuming copy_q == False
return Normal(qz3 * 1, scale=1)
 where
  qz3 = Normal(z2 * 2, scale=1)
  z2 = Normal(z1 * 1, scale=1)
  z1 = Normal(z0 * 1, scale=1)
```

What happens if we set `copy_q=True`? Let's return to the point where copy was looking at the arguments of `zs[4]` and it reached `zs[3]`. Now, in this new scenario, the arguments of `zs[4]` will no longer contain `zs[3]` directly, because `zs[3]` is first multiplied with `phi`. So the first argument will be the result of that multiplication:

```python
In [2]: zs[4]._kwargs
Out[2]: 
{'loc': <tf.Tensor 'mul_4:0' shape=(1,) dtype=float32>,
 'name': 'z4/',
 'scale': 1.0}
```

This `tf.Tensor` will then direct us to the `tf.Operation` that created it and so on... To keep things simple, I'm going to assume that this will just indirectly lead us to `zs[3]` without any other side-effects, and I'm going to ignore this indirect relationship in the rest of my thought process.

Since `zs[3]` is paired up with `qzs[3]`, copy will turn its gaze towards `qzs[3]`, as can be seen in:

```python
if org_instance in dict_swap and replace_itself:  # replace_itself==True
  org_instance = dict_swap[org_instance]
  if not copy_q:
    return org_instance
```

Here `replace_itself` is `True` because it was set to `True` explicitly by the recursive call. Now, this means that the next step of recursion will be into the arguments of `qzs[3]`, which is `zs[2]`. At this point, once again copy detects that `zs[2]` is paired with `qzs[2]` and turns its attention to `qzs[2]`. This goes on until we reach `zs[0]`, where for the last time we find its pair `qzs[0]` and we climb back down the stack trace, at each step instantiating a new random variable by copying and replacing the arguments of the old one, starting with `qzs[0]`. `qzs[0]` does not have any arguments depending on further random variables. So `copy_zs0` will be created by taking the arguments of `qzs[0]`:

```python
In [25]: qzs[0]._kwargs
Out[25]: 
{'loc': <tf.Tensor 'zeros:0' shape=(1,) dtype=float32>,
 'name': 'qz0/',
 'scale': 1.0}
```

and copying them and using the copies to call the constructor to `Normal`. The return value of this constructor call will be the copy of `zs[0]`. We can inspect its arguments:

```python
In [27]: copy_zs0._kwargs
Out[27]: 
{'loc': <tf.Tensor 'copy_z0/zeros:0' shape=(1,) dtype=float32>,
 'name': 'copy_z0/qz0/',
 'scale': 1.0}

In [28]: id(copy_zs0._kwargs['loc'])
Out[28]: 140117468395784

In [29]: id(qzs[0]._kwargs['loc'])
Out[29]: 140117784798656
```

And `copy` will keep on copying in the same way until reaching the top of the stack trace, which is `copy(zs[4], latent_vars, copy_q=True)` where we were trying to copy the arguments of `zs[4]`. Mind you, the arguments of `zs[4]`, not the arguments of `qzs[4]`. This is the only part of the whole copying process where copy did not turn its gaze from a `z` to a `qz` and instead went on to copy the arguments of the `z`. This is because the original call `copy(zs[4], latent_vars, copy_q=True)` did not set `replace_itself=True`. As a consequence, the result we will get is going to be this:

```
# assuming copy_q == True
return Normal(qz3 * 1, scale=1)
 where
  qz3 = Normal(qz2 * 2, scale=1)
  qz2 = Normal(qz1 * 2, scale=1)
  qz1 = Normal(qz0 * 2, scale=1)
```

Hence at all steps but the last one, we take downward arrows and at the last step we erroneously take a sideways arrow. To prevent this, we need to change the original call to: `copy(zs[4], latent_vars, copy_q=True, replace_itself=True)`.
Now, at the very first call to `copy`, `qzs[4]` will be detected as the pair of `zs[4]` and the recursion will be into the arguments of `qzs[4]`, not the arguments of `zs[4]`.
Finally let's compare the three different function calls in the setting of the modified toy example:

```python
copy(zs[4], latent_vars)
copy(zs[4], latent_vars, copy_q=True)
copy(zs[4], latent_vars, replace_itself=True, copy_q=True)
In [5]: copy_zs4_1 = ed.copy(zs[4], latent_vars, scope='test1')

In [6]: copy_zs4_2 = ed.copy(zs[4], latent_vars, scope='test2', copy_q=True)

In [7]: copy_zs4_3 = ed.copy(zs[4], latent_vars, scope='test3', replace_itself=True, copy_q=True)
```

We've made three copies under different scopes. Now let's look at the ancestors:

```python
In [9]: copy_zs4_1.get_ancestors()
Out[9]: 
[<ed.RandomVariable 'qz3/' shape=(1,) dtype=float32>,
 <ed.RandomVariable 'z2/' shape=(1,) dtype=float32>,
 <ed.RandomVariable 'z1/' shape=(1,) dtype=float32>,
 <ed.RandomVariable 'z0/' shape=(1,) dtype=float32>]

In [10]: copy_zs4_2.get_ancestors()
Out[10]: 
[<ed.RandomVariable 'test2/qz3/' shape=(1,) dtype=float32>,
 <ed.RandomVariable 'test2/qz0/' shape=(1,) dtype=float32>,
 <ed.RandomVariable 'test2/qz1/' shape=(1,) dtype=float32>,
 <ed.RandomVariable 'test2/qz2/' shape=(1,) dtype=float32>]

In [11]: copy_zs4_3.get_ancestors()
Out[11]: 
[<ed.RandomVariable 'test3/qz3/' shape=(1,) dtype=float32>,
 <ed.RandomVariable 'test3/qz0/' shape=(1,) dtype=float32>,
 <ed.RandomVariable 'test3/qz1/' shape=(1,) dtype=float32>,
 <ed.RandomVariable 'test3/qz2/' shape=(1,) dtype=float32>]
```

The first call gets the ancestor relationship wrong. The second and third don't seem to have this problem. However, this alone doesn't prove that everything is in order. We also have to look at the functional dependency between parents and children. I'll only look at how the `loc` argument of `copy_zs4_x` is obtained given its parent:

```python
In [12]: copy_zs4_1._kwargs['loc'].op
Out[12]: <tf.Operation 'test1/mul_4' type=Mul>

In [13]: copy_zs4_2._kwargs['loc'].op
Out[13]: <tf.Operation 'test2/mul_4' type=Mul>

In [14]: copy_zs4_3._kwargs['loc'].op
Out[14]: <tf.Operation 'test3/mul_9' type=Mul>
```

It seems that the copies produced by the first two calls both receive their `loc` argument from a multiplication called `mul_4`, whereas in the third function call, the `loc` is coming from another operation called `mul_9`.

**Comment by dustinvtran** [comment]
Update: This issue will be on hold (so no need to work on it). I'm in the midst of a redesign for Edward 2.0. It involves no graph copying.
